Click to filter values in Discover doc table and Visualize data table

Bargs opened the pull request:
Fixes #3432
Replaces https://github.com/elastic/kibana/pull/5344

This PR extends the filter icons implemented in #5344 to the data table visualization.

![screen shot 2017-01-20 at 5 59 17 pm](https://cloud.githubusercontent.com/assets/6239176/22168389/484ddd8a-df3a-11e6-8e76-e152ef280f1a.png)

lukasolson commented on src/ui/public/directives/rows.js:
```diff
@@ -18,15 +18,33 @@ module.directive('kbnRows', function ($compile, $rootScope, getAppState, Private
+                ></span>
+              </span>
+            </td>`
+          );
```
Could you move this into its own template file?

lukasolson commented on src/ui/public/directives/rows.js:
```diff
@@ -18,15 +18,33 @@ module.directive('kbnRows', function ($compile, $rootScope, getAppState, Private
+          $cell.scope.clickHandler = function (event, negate) {
             if ($(event.target).is('a')) return; // Don't add filter if a link was clicked
-            clickHandler({ point: { aggConfigResult: aggConfigResult } });
+            clickHandler({ point: { aggConfigResult: aggConfigResult }, negate: negate });
```
The [style guide](https://github.com/elastic/kibana/blob/fd6fd6411e2fa4aff8bec098943f8bac31d70163/style_guides/js_style_guide.md) suggests using the property value shorthand syntax (simply `negate` instead of `negate: negate`).

jimmyjones2 commented:
This is awsome @Bargs, thanks!

Two minor inconsistencies between discover and the data table - whole row in data table is highlighted on mouseover and magnifying glass shown at the top of the row.

Bargs commented:
Thanks for the input @jimmyjones2! The cell/row highlighting was an existing style of the data table. Perhaps you're right we should remove it now. What do you think @cjcenizal?

Could you provide a screenshot of the positioning inconsistency you mentioned? From what I could tell the doc table also put the magnifying icon at the top of the row but I may have missed something.

jimmyjones2 commented:
@Bargs Sorry I wasn't very clear - in the header of the data table there is a magnifying glass as a cue for filtering, but not in the doc table.

cjcenizal commented:
This looks good from a UX perspective, except for one weird thing I found. If I tried to filter on a value, sometimes it tried to apply a filter for every value in the _row_, instead of just the value of the cell I clicked on. When this happened, it prompted me to verify that I want to filter. If I selected a different cell, this behavior didn't occur, and instead the filter was applied for just the cell I clicked on. From a UX perspective, I couldn't figure out why this was happening. Any idea what's going on here?

![data_table_filters](https://cloud.githubusercontent.com/assets/1238659/22222537/17043114-e16c-11e6-923d-17661deecd9c.gif)

cjcenizal commented on src/ui/public/doc_table/components/table_row.js:
```diff
@@ -83,27 +85,44 @@ module.directive('kbnTableRow', function ($compile) {
-            formatted: _displayField(row, column, true)
+            formatted: _displayField(row, column, true),
+            filterable: isFilterable,
+            column: column
```
This can also just be `column`, similar to @lukasolson's comment.

Bargs commented:
@cjcenizal this is the way the current filtering works in the data table ¯\_(ツ)_/¯

cjcenizal commented:
@Bargs Gotcha thanks. I dug into this a bit and created a ticket for improving this UX: https://github.com/elastic/kibana/issues/10024

I agree that we should remove the hover highlight, since we no longer need it to signify that the cell is clickable.

ppisljar commented:
use hover to highlight whole row then ?

cjcenizal commented:
@ppisljar I'm open to trying that out. What do you think the biggest benefit(s) of that is for the user?

ppisljar commented:
with multiple columns its hard to see which value matches to the row you are looking to, highlighting the row makes it obvious.

Bargs commented:
@lukasolson @cjcenizal @ppisljar Just pushed some updates:

* Data table now only highlights row on hover, not individual cell
* Fixed filter icon colors in dark theme
* Extracted filter html to a template
* Use object shorthand in a couple of places
* Removed outdated filter tool tip at the top of data table columns

Please take another look.

Bargs commented:
jenkins, test this

lukasolson commented:
Hmm, how complicated would it be to centralize this logic into a directive and re-use it in both places? That way we wouldn't have the HTML duplicated in two different places... What do you think?

Bargs commented:
@lukasolson I think a bit of it could be DRYed up, but not a ton. The way the filters are built is quite different in each context, so it would really just be the html that gets de-duplicated. I don't mind doing it, but I also know @ppisljar is working on some major changes to the data table so it might be best to avoid sharing this logic for now.

marenas commented on src/ui/public/doc_table/components/table_row.js:
```diff
@@ -128,7 +147,7 @@ module.directive('kbnTableRow', function ($compile) {
-          if (i === 0 && !reuse) {
+          if (!reuse) {
             $toggleScope = $scope.$new();
             $compile($target)($toggleScope);
```
Sorry for the late and random comment, but doesn't this `$compile($target)($toggleScope);` on all "newHtmls" elements for each row introduce some performance degradation when showing many rows ?

Bargs replied:
@marenas I haven't personally seen any slowdown with this change. The template for each cell is pretty minimal.

marenas replied:
makes sense, thanks !
